### ISSUE TITLE
Running intermediate try scripts with the original shell

### ISSUE THREAD

**Issue by angelhof** [binpash/try#7]
`try` currently runs the intermediate temporary scripts that it creates using `/bin/sh`. This causes a problem if we want the internal script to run using a different shell and therefore inherit its state, e.g., bash functions. 

A running example of the wanted behavior follows:
```sh
$ func() { echo hi; }
$ export -f func
$ bash ./deps/try/try func
hi
```

I think that a good portable solution would be to run the intermediate scripts with whatever shell try is running too. What is a portable way to determine which shell we are running on? Here are some alternatives:
- The `$SHELL` variable will contain the name of the login shell (it doesn't work if the subprocess is different than the login shell)
- Using `ps`, e.g., `ps -o args= -p "$$"` or `ps -o command= -p "$$"`. This seems to me like the most portable and correct solution. 

Here are some SO posts that discuss this issue: 
- https://stackoverflow.com/questions/3327013/how-to-determine-the-current-interactive-shell-that-im-in-command-line
- https://unix.stackexchange.com/questions/71121/determine-shell-in-script-during-runtime

Note that we don't really care about the underlying shell, we just want the executable (or even a link to it), so if try was invoked using `/bin/sh` we can just use that, we don't care that it was bash under the hood.

**Comment by angelhof** [comment]
@mgree and @dspinellis any thoughts? I can implement the `ps` solution, I just need a second opinion on whether it is adequately portable :)

**Comment by mgree** [comment]
I agree that `ps` doesn't quite cut it. These detection approaches seem good; we should also have a flag that lets you pick the shell that runs.

Also: these APIs only exist on Linux and FreeBSD, so we don't need to worry about macOS portability.

**Comment by kwakubiney** [comment]
@mgree which APIs do you speak of? Are you referring to the APIs exposed by `try`?

**Comment by mgree** [comment]
I'm talking about `unshare`.

**Comment by mgree** [comment]
I propose a three-tiered approach:

- If the `$TRY_SHELL` variable is set, use that. (E.g., `TRY_SHELL=/usr/bin/bash5.3 try ...`). (Could be a command-line argument instead? Dunno.)
- If the shell running try is not exactly `/bin/sh`, we should use that shell. (Look in `/proc/$$/exe`.)
- If the parent process of the try `/bin/sh` process is a particular shell, we should use that. (Look in `/proc/$PPID/exe`.)
- If `$SHELL` is specified and neither of the above holds, use that.

**Comment by mgree** [comment]
Closed by #197
